### PR TITLE
Fix resource leak

### PR DESCRIPTION
**Pull request by FaramosCZ**

### CODE REVIEW

**Comment by FaramosCZ** [comment]
Submitted under MCA

---

Fix:
https://github.com/MariaDB/server/pull/983
https://github.com/MariaDB/server/pull/984

**Comment by FaramosCZ** [comment]
--

Fix for several cwe-772 found by covscan:

```
Error: RESOURCE_LEAK (CWE-772):
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5199: alloc_fn: Storage is returned from allocation function "os_file_opendir".
mariadb-10.3.10/storage/innobase/os/os0file.cc:2780:2: alloc_fn: Storage is returned from allocation function "opendir".
mariadb-10.3.10/storage/innobase/os/os0file.cc:2780:2: var_assign: Assigning: "dir" = "opendir(dirname)".
mariadb-10.3.10/storage/innobase/os/os0file.cc:2782:2: cond_false: Condition "dir == NULL", taking false branch.
mariadb-10.3.10/storage/innobase/os/os0file.cc:2784:2: if_end: End of if statement.
mariadb-10.3.10/storage/innobase/os/os0file.cc:2786:2: return_alloc: Returning allocated memory "dir".
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5199: var_assign: Assigning: "dbdir" = storage returned from "os_file_opendir(path, false)".
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5201: cond_true: Condition "dbdir != NULL", taking true branch.
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5202: noescape: Resource "dbdir" is not freed or pointed-to in "fil_file_readdir_next_file".
mariadb-10.3.10/storage/innobase/fil/fil0fil.cc:3988:16: noescape: "fil_file_readdir_next_file(dberr_t *, char const *, os_file_dir_t, os_file_stat_t *)" does not free or save its parameter "dir".
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5204: cond_true: Condition "ret == 0", taking true branch.
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5205: cond_false: Condition "fileinfo.type == OS_FILE_TYPE_DIR", taking false branch.
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5207: if_end: End of if statement.
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5209: cond_true: Condition "strlen(fileinfo.name) > suffix_len", taking true branch.
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5209: cond_true: Condition "0 == strcmp(fileinfo.name + strlen(fileinfo.name) - suffix_len, suffix)", taking true branch.
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5213: cond_true: Condition "!(*func)(path, NULL, fileinfo.name, NULL)", taking true branch.
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5217: leaked_storage: Variable "dbdir" going out of scope leaks the storage it points to.
# 5215|                                             fileinfo.name, NULL))
# 5216|                                 {
# 5217|->                                       return(FALSE);
# 5218|                                 }
# 5219|                         }
```

```
Error: RESOURCE_LEAK (CWE-772):
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5253: alloc_fn: Storage is returned from allocation function "os_file_opendir".
mariadb-10.3.10/storage/innobase/os/os0file.cc:2780:2: alloc_fn: Storage is returned from allocation function "opendir".
mariadb-10.3.10/storage/innobase/os/os0file.cc:2780:2: var_assign: Assigning: "dir" = "opendir(dirname)".
mariadb-10.3.10/storage/innobase/os/os0file.cc:2782:2: cond_false: Condition "dir == NULL", taking false branch.
mariadb-10.3.10/storage/innobase/os/os0file.cc:2784:2: if_end: End of if statement.
mariadb-10.3.10/storage/innobase/os/os0file.cc:2786:2: return_alloc: Returning allocated memory "dir".
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5253: var_assign: Assigning: "dbdir" = storage returned from "os_file_opendir(dbpath, false)".
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5255: cond_true: Condition "dbdir != NULL", taking true branch.
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5257: noescape: Resource "dbdir" is not freed or pointed-to in "fil_file_readdir_next_file".
mariadb-10.3.10/storage/innobase/fil/fil0fil.cc:3988:16: noescape: "fil_file_readdir_next_file(dberr_t *, char const *, os_file_dir_t, os_file_stat_t *)" does not free or save its parameter "dir".
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5259: cond_true: Condition "ret == 0", taking true branch.
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5261: cond_false: Condition "fileinfo.type == OS_FILE_TYPE_DIR", taking false branch.
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5264: if_end: End of if statement.
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5266: cond_true: Condition "strlen(fileinfo.name) > suffix_len", taking true branch.
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5266: cond_true: Condition "0 == strcmp(fileinfo.name + strlen(fileinfo.name) - suffix_len, suffix)", taking true branch.
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5273: cond_true: Condition "!(*func)(path, dbinfo.name, fileinfo.name, NULL)", taking true branch.
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5278: leaked_storage: Variable "dbdir" going out of scope leaks the storage it points to.
# 5276|                                                     fileinfo.name, NULL))
# 5277|                                         {
# 5278|->                                               return(FALSE);
# 5279|                                         }
# 5280|                                 }
```


```
Error: RESOURCE_LEAK (CWE-772):
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5233: alloc_fn: Storage is returned from allocation function "os_file_opendir".
mariadb-10.3.10/storage/innobase/os/os0file.cc:2780:2: alloc_fn: Storage is returned from allocation function "opendir".
mariadb-10.3.10/storage/innobase/os/os0file.cc:2780:2: var_assign: Assigning: "dir" = "opendir(dirname)".
mariadb-10.3.10/storage/innobase/os/os0file.cc:2786:2: return_alloc: Returning allocated memory "dir".
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5233: var_assign: Assigning: "dir" = storage returned from "os_file_opendir(path, false)".
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5240: noescape: Resource "dir" is not freed or pointed-to in "fil_file_readdir_next_file".
mariadb-10.3.10/storage/innobase/fil/fil0fil.cc:3988:16: noescape: "fil_file_readdir_next_file(dberr_t *, char const *, os_file_dir_t, os_file_stat_t *)" does not free or save its parameter "dir".
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5290: noescape: Resource "dir" is not freed or pointed-to in "fil_file_readdir_next_file".
mariadb-10.3.10/storage/innobase/fil/fil0fil.cc:3988:16: noescape: "fil_file_readdir_next_file(dberr_t *, char const *, os_file_dir_t, os_file_stat_t *)" does not free or save its parameter "dir".
mariadb-10.3.10/extra/mariabackup/xtrabackup.cc:5278: leaked_storage: Variable "dir" going out of scope leaks the storage it points to.
# 5276|                                                     fileinfo.name, NULL))
# 5277|                                         {
# 5278|->                                               return(FALSE);
# 5279|                                         }
# 5280|                                 }
```

**Comment by shinnok** [comment]
@FaramosCZ Hi, you mentioned in a different PR that you are using Coverity for digging these out? I am trying to set this up too so that we can have Coverity reports automated, though my set up didn't yield much result. Can you please share your setup with me or jump on Zulip where we hang out and chat, so I can pick your brain; well your Coverity setup :)
https://mariadb.zulipchat.com/

File: extra/mariabackup/xtrabackup.cc [hunk 1]
```diff
@@ -5295,6 +5296,7 @@ xb_process_datadir(
 						    dbinfo.name,
 						    fileinfo.name, data))
 					{
+						os_file_closedir(dbdir);
 						return(FALSE);
```
Comment (dr-m) [inline extra/mariabackup/xtrabackup.cc]: I think that we must close also `dir` here.

**Comment by an3l** [comment]
Can you please fix the commit message ?

**Comment by FaramosCZ** [comment]
> Can you please fix the commit message ?

Should be foxed ... I mean fixed :)

Thanks for the catch.

**Comment by svoj** [comment]
Thanks! 2/3 of the fix are in 10.1 already: 7372fe4da195f465a2f13984dcd89325ec361c43. Merged `os_file_closedir(dir);` as 9c9bf9642e5693841375350540530141e1916b36.

**Pull request closed without merging**